baseplate/context/hvac.py

```diff
@@ -0,0 +1,119 @@
+import requests
+
+from . import ContextFactory
+from .. import config
+
+
+def hvac_factory_from_config(app_config, secrets_store, prefix="vault."):
+    """Make an HVAC client factory from a configuration dictionary.
+
+    The keys useful to :py:func:`hvac_factory_from_config` should be prefixed,
+    e.g.  ``vault.url``, ``vault.timeout``, etc. The ``prefix`` argument
+    specifies the prefix used to filter keys.
+
+    Supported keys:
+
+    * ``url`` (required): URL to find the Vault cluster at.
+    * ``timeout``: How long to wait for calls to Vault.
+
+    :param dict app_config: The raw application configuration.
+    :param baseplate.secrets.SecretsStore secrets_store: A configured secrets
+        store from which we can get a Vault authentication token.
+    :param str prefix: The prefix for configuration keys.
+
+    """
+    assert prefix.endswith(".")
+    config_prefix = prefix[:-1]
+    cfg = config.parse_config(app_config, {
+        config_prefix: {
+            "url": config.String,
+            "timeout": config.Optional(config.Timespan,
+                                       default=datetime.timedelta(seconds=1)),
+        },
+    })
+    options = getattr(cfg, config_prefix)
+
+    return HvacContextFactory(
+        options.url,
+        secrets_store,
+        options.timeout,
+    )
+
+
+class HvacContextFactory(ContextFactory):
+    """HVAC client context factory.
+
+    This factory will attach a proxy object which acts like an
+    :py:class:`hvac.Client` to an attribute on the :term:`context object`. All
+    methods that talk to Vault will be automatically instrumented for tracing
+    and diagnostic metrics.
+
+    :param str url: URL to find the Vault cluster at.
+    :param baseplate.secrets.SecretsStore secrets_store: Configured secrets
+        store from which we can get a Vault authentication token.
+    :param datetime.timedelta timeout: How long to wait for calls to Vault.
+
+    """
+    def __init__(self, url, secrets_store, timeout):
+        self.url = url
+        self.secrets = secrets_store
+        self.timeout = timeout
+        self.session = requests.Session()
+
+    def make_object_for_context(self, name, server_span):
+        vault_token = self.secrets.get_vault_token()
+
+        return InstrumentedHvacClient(
+            url=self.url,
+            token=vault_token,
+            timeout=self.timeout.total_seconds(),
+            session=self.session,
+            context_name=name,
+            server_span=server_span,
+        )
+
+
+class InstrumentedHvacClient(hvac.Client):
+    def __init__(self, url, token, timeout, session, context_name, server_span):
+        self.context_name = context_name
+        self.server_span = server_span
+
+        super(InstrumentedHvacClient, self).__init__(
+            url=url, token=token, timeout=timeout, session=session)
+
+    # this ugliness is us undoing the name mangling that __request turns into
+    # inside python. this feels very dirty.
```

Comment:
What if instead of instrumenting the hvac client itself, we instrumented the Session object?